.github/workflows/pull-request.yml

```diff
@@ -0,0 +1,169 @@
+        run: |
+          sudo snap install yq
+          curl ${{ env.HELM_URL_V3 }} | tar xvzf - --strip-components=1 -C /tmp/ && \
+          sudo mv /tmp/helm /usr/bin/helm_v3 && \
+          sudo chmod +x /usr/bin/helm_v3
+          helm_v3 plugin install https://github.com/helm-unittest/helm-unittest.git --version ${{ env.HELM_UNITTEST_VERSION }}; \
+      - name: Build
+        run: ./scripts/chart/build chart
+      - name: Validate
+        run: ./scripts/chart/validate
+      - name: Test
+        run: ./scripts/chart/test
+      - name: Package
+        run: ./scripts/chart/package
+  build-server:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [linux]
+        arch: [amd64, arm64]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Setup Environment Variables
+        run: |
+          echo "ARCH=${{ matrix.arch }}" >> "$GITHUB_ENV"
+          echo "RKE_VERSION=$(grep -m1 'github.com/rancher/rke' go.mod | awk '{print $2}')" >> "$GITHUB_ENV"
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ env.IMAGE }}
+          flavor: |
+            latest=false
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+      - name: Create k3s images file
+        run: touch k3s-airgap-images.tar
+      - name: Download data.json
+        run: curl -sLf https://releases.rancher.com/kontainer-driver-metadata/${{ env.CATTLE_KDM_BRANCH }}/data.json > ./data.json
+      - name: Build and export server
+        id: build
+        uses: docker/build-push-action@v5
+        with:
+          push: false
+          build-args: |
+            "VERSION=${{ env.COMMIT }}"
+            "COMMIT=${{ env.COMMIT }}"
+            "RKE_VERSION=${{ env.RKE_VERSION }}"
+            "ARCH=${{ matrix.arch }}"
+          tags: ${{ env.IMAGE }}:${{ env.TAG }}-${{ matrix.arch }}
+          context: .
+          platforms: "${{ matrix.os }}/${{ matrix.arch }}"
+          file: ./package/Dockerfile
+          labels: "${{ steps.meta.outputs.labels }}"
+          outputs: type=docker,dest=/tmp/rancher-${{ matrix.os }}-${{ matrix.arch }}.tar
+      - name: Upload image
+        uses: actions/upload-artifact@v4
+        with:
+          name: "rancher-${{ matrix.os }}-${{ matrix.arch }}"
+          path: /tmp/rancher-${{ matrix.os }}-${{ matrix.arch }}.tar
+          if-no-files-found: error
+          retention-days: 1
+          overwrite: false
+  build-agent:
+    needs: [build-server]
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [linux]
+        arch: [amd64, arm64]
+    services:
+      registry:
+        image: registry:2
+        ports:
+          - 5000:5000
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+      - name: Clean runner
```

Comment:
Valid point!

@tashima42 and what about cleaning the Go cache too? WDYT?

Reply:
It isn't a problem now, but it could be in the future, so it's a good idea to also add it to `v2.8`. When we have the final `v2.9` version, I'll backport everything.
About cache, I think it wouldn't be a problem using GH runners, but there's no harm on adding it and the upside of preventing the same problems as we had before.